From d37459dd9ac6ee93f5e4973ad727d0313420f8aa Mon Sep 17 00:00:00 2001
From: Matthew Wellings <matthew.wellings@arm.com>
Date: Wed, 27 Nov 2024 15:23:47 +0000
Subject: [PATCH] framework: Add fwk_str_is_in_boundry function

This change adds an additional function, fwk_str_is_in_boundry, that can
be used to check if a copy will be within the bounds of the destination
buffer.

This is a simple maths function, returns true if the copy will be in
bounds otherwise returns false.

Also adds associated tests.

Signed-off-by: Matthew Wellings <matthew.wellings@arm.com>
Upstream-Status: Backport [https://gitlab.arm.com/firmware/SCP-firmware/-/commit/25eaaf9b59053f740121772f563ca5ec0b825848]
Signed-off-by: Debbie Martin <Debbie.Martin@arm.com>
---
 framework/include/fwk_string.h   | 18 +++++++++++-
 framework/src/fwk_string.c       |  9 ++++++
 framework/test/test_fwk_string.c | 47 +++++++++++++++++++++++++++++++-
 3 files changed, 72 insertions(+), 2 deletions(-)

diff --git a/framework/include/fwk_string.h b/framework/include/fwk_string.h
index 2aff8e41b..aa4842093 100644
--- a/framework/include/fwk_string.h
+++ b/framework/include/fwk_string.h
@@ -1,6 +1,6 @@
 /*
  * Arm SCP/MCP Software
- * Copyright (c) 2021, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2021-2024, Arm Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -10,6 +10,7 @@
 
 #include <fwk_attributes.h>
 
+#include <stdbool.h>
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
@@ -92,6 +93,21 @@ FWK_LEAF FWK_NOTHROW void fwk_str_strncpy(
     const char *src,
     size_t count);
 
+/*!
+ * \brief Checks that a copy will be within the bounds of the destination buffer
+ *
+ * \details This is a simple maths function, returns true if the copy will be in
+ *      bounds otherwise returns false.
+ *
+ * \param[in] buffer_size Size of a buffer to copy to
+ * \param[in] offset Offset of copy within destination buffer
+ * \param[in] data_size number of bytes to copy
+ */
+FWK_LEAF FWK_NOTHROW bool fwk_str_is_in_boundry(
+    size_t buffer_size,
+    size_t offset,
+    size_t data_size);
+
 /*!
  * \}
  */
diff --git a/framework/src/fwk_string.c b/framework/src/fwk_string.c
index 65aa69556..977a43a6c 100644
--- a/framework/src/fwk_string.c
+++ b/framework/src/fwk_string.c
@@ -54,3 +54,12 @@ void fwk_str_strncpy(char *dest, const char *src, size_t count)
         fwk_trap();
     }
 }
+
+bool fwk_str_is_in_boundry(size_t buffer_size, size_t offset, size_t data_size)
+{
+    if ((offset > buffer_size) || (data_size > (buffer_size - offset))) {
+        return false;
+    }
+
+    return true;
+}
diff --git a/framework/test/test_fwk_string.c b/framework/test/test_fwk_string.c
index eeb6d47a4..9327eaaeb 100644
--- a/framework/test/test_fwk_string.c
+++ b/framework/test/test_fwk_string.c
@@ -1,6 +1,6 @@
 /*
  * Arm SCP/MCP Software
- * Copyright (c) 2021, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2021-2024, Arm Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -52,10 +52,55 @@ static void test_fwk_str_strncpy(void)
     }
 }
 
+static void test_fwk_str_is_in_boundry_invalid_param_large_size(void)
+{
+    size_t offset = 50;
+    size_t buffer_size = 100;
+
+    bool result =
+        fwk_str_is_in_boundry(buffer_size, offset, SIZE_MAX - (offset / 2));
+
+    assert(result == false);
+}
+
+static void test_fwk_str_is_in_boundry_invalid_param_large_offset(void)
+{
+    size_t buffer_size = 100;
+    size_t size = 5;
+
+    bool result = fwk_str_is_in_boundry(buffer_size, SIZE_MAX - 2, size);
+
+    assert(result == false);
+}
+
+static void test_fwk_str_is_in_boundry_invalid_param_beyond_end(void)
+{
+    size_t buffer_size = 100;
+    size_t size = 1;
+
+    bool result = fwk_str_is_in_boundry(buffer_size, buffer_size, size);
+
+    assert(result == false);
+}
+
+static void test_fwk_str_is_in_boundry_valid_param_before_end(void)
+{
+    size_t buffer_size = 100;
+    size_t size = 1;
+
+    bool result = fwk_str_is_in_boundry(buffer_size, buffer_size - 1, size);
+
+    assert(result == true);
+}
+
 static const struct fwk_test_case_desc test_case_table[] = {
     FWK_TEST_CASE(test_fwk_str_memset),
     FWK_TEST_CASE(test_fwk_str_memcpy),
     FWK_TEST_CASE(test_fwk_str_strncpy),
+    FWK_TEST_CASE(test_fwk_str_is_in_boundry_invalid_param_large_size),
+    FWK_TEST_CASE(test_fwk_str_is_in_boundry_invalid_param_large_offset),
+    FWK_TEST_CASE(test_fwk_str_is_in_boundry_invalid_param_beyond_end),
+    FWK_TEST_CASE(test_fwk_str_is_in_boundry_valid_param_before_end),
 };
 
 struct fwk_test_suite_desc test_suite = {
-- 
2.25.1

